### ISSUE TITLE
Add check to prohibit Windows line endings

### ISSUE THREAD

**Issue by gaul** [checkstyle/checkstyle#893]
While `RegexpMultiline` can do this it is not discoverable by users:

``` xml
<module name="RegexpMultiline">
    <property name="format" value="\r\n"/>
    <property name="message" value="Do not use Windows line endings"/>
</module>
```
## 

ATTENTION: new Check discussion is moved to new issue - https://github.com/checkstyle/checkstyle/issues/18109

**Comment by romani** [comment]
We always try to make a general Checks, so check should enforce user defined LineEnd symbol.

We will consider that idea as soon as we finish moratorium period for new Checks, if you ready to implement it , please be welcome at our experimental Check project - https://github.com/sevntu-checkstyle/sevntu.checkstyle.

**Comment by pbaranchikov** [comment]
This check is completely useless, as good VCS configuration will force local working copy to have Windows-like (CRLF) EOL-style on Windows clients and Linux EOL-style (LF) on Linux clients. In this way, the correctly configured clients should have multiple errors for checkstyle.

This contstraint is to be forced inside your VCS. For Stash you may use https://github.com/pbaranchikov/stash-eol-check, for gitolite-admin you should perform some BASH-scripting and so on.

**Comment by romani** [comment]
@andrewgaul , what kind of VCS you have ? GIT manage it as pbaranchikov described.

**Comment by andrewgaul** [comment]
@pbaranchikov I agree that correctly configured git clients can warn about line endings via various hooks, although as a project admin using GitHub there is no real way to enforce this on either the server or our contributors' clients.  We enabled previously mentioned `RegexpMultiline` for jclouds/jclouds#717.

**Comment by mkordas** [comment]
I don't think this rule is a good idea. Using git, you have no control which EOL style your clients use. We may end up with rule that passes on CI, but fails for people using Windows.

**Comment by isopov** [comment]
At my work we are using git without enforcing line endings and we do enforce them with checkstyle in java sources. 
Test inputs have different line endings to test that file based logic works as expected with any combination.

**Comment by msteiger** [comment]
I completely agree with @andrewgaul - such a check could be very valuable imho. Inconsistent line endings can certainly cause problems, in particular for resources files (config files, csv files, etc).

Let's assume for a second that *git* is used. Clients can decide how files are converted using the [core.autocrlf feature](https://help.github.com/articles/dealing-with-line-endings/). Some of them might have turned it off, leading to Windows clients pushing `CRLF` line endings, which is not supposed to happen according to git standards. If your code assumes `LF` line endings, a CSV file might become unreadable :cry: 

@mkordas Actually, you can configure line endings very nicely using the `.gitattributes` file. See #1045 for example. You can also specify line endings per file type and git can also try to automatically detect and performs EOL conversion. Unfortunately, the jgit implementation that eclipse uses simply [ignores the `.gitattributes` file](https://bugs.eclipse.org/bugs/show_bug.cgi?id=342372) :cry: 

**To sum it up:** A check that verifies that only text/code files with `LF` line endings are committed makes perfectly sense to me as ensures consistency of the code files and thus helps avoiding bugs that are really hard to track down.

**Comment by mkordas** [comment]
@msteiger, generally I agree with you. I had issues with newlines in my projects too. I have just couple concerns:
1. Let's assume we have developers using Windows and CI on Linux and check is configured to enforce `LF` - Checkstyle would fail on development machines
2. Let's assume we've added property e.g. "match system line separator" to enforce `LF` on Unix and `CRLF` on Windows - Checkstyle would fail for all developers that have `autocrlf` disabled
3. Let's assume we use `.gitattributes` file - then each entry from this file needs suppression in Checkstyle config...
4. Let's assume developer has all files as `CRLF` on Windows but temporarily wants to execute bash script using Cygwin, so `dos2unix` is required. Checkstyle would fail, even if then Git afterwards would manage that case.

I'm just not convinced to have the check that is so fragile, but maybe just good description (and warning) in documentation would be enough.

**Comment by vorburger** [comment]
+1 @andrewgaul et al IMHO this is most certainly a good idea and would be great to have built-in! Note "(?s:\r\n.*)" is better because it will only match the first wrong newline, so less spammy logs (and perhaps even faster?). See also my blurb (blog post) about this on http://blog2.vorburger.ch/2015/06/eol.html.

**Comment by romani** [comment]
There are bunch of VCS, so might be some VCS does not handle that correctly.

Somebody could make this Check if it is required, all are welcome to contribute.

**Comment by vorburger** [comment]
Interesting related problem came up in https://github.com/checkstyle/checkstyle/pull/3557:

If you are in a multi-OS environment, and you use core.autocrlf, then e.g. when running on Windows git will transform LF to CR+LF on disk, right? So this kind of regexp check would fail there; actually has failed in Checkstyle's own build, some of which running on Windows; see https://github.com/checkstyle/checkstyle/pull/3557. 

So if such a new Check was developed it would have to have some sort of mode like "this project is part of a repo using core.autocrlf, and so if running on Windows OS then skip this check" - thoughts, anyone?

Or perhaps using core.autocrlf is just a bad idea? ;-) Start by removing it from CS's own repo?

**Comment by rnveach** [comment]
> perhaps using core.autocrlf and having Windows build machines is just a bad idea?

Businesses and private users are the ones who really stick to using Windows. Most businesses may not use Windows on production build machines, but they still need to be able to test them locally.
Abandoning windows from any project will diminish your user base, which is usually a bad idea.

> repo using core.autocrlf

You would also need to take into account `.gitattributes` which can force different line endings for specific files.

> repo using core.autocrlf, and so if running on Windows OS then skip this check

Nothing prevents Linux users from turning option on. Maybe repo has legacy code and was developed and solely used by windows. You would have to skip all who use `autocrlf`.

> if such a new Check was developed it would have to have some sort of mode like "this project is part of a repo using core.autocrlf, and so if running on Windows OS then skip this check" 

Why is it an issue to examine local copy (Checkstyle's domain) and not examine the server's copy (git's domain)? Checkstyle may not be set to validate every file in repository too.
This sounds to me more like an issue not validating the commits going into the repository than the files on the local hard drive.

**Comment by gdemecki** [comment]
I definitely agree  @vorburger  and @msteiger. Such a check would be valuable.

Although there may exists some projects which need CRLF line endings - I believe more than 9/10 should stick with the LF and `core.autocrlf = input` (on every platform). I've seen far too many accidental CRLF endings in my life.

@mkordas Please note that all 4 points raised by you are solved with a single Git option: `autocrlf = input`.

And all who still disagree, please remember that this module, as every other, should be enabled when project would need this - **nobody is forced to do so**.

**Comment by mkordas** [comment]
@gdemecki Now it's more than year after my comment from above and I had so many fights with incorrect line endings in projects that I definitely see a need for such check. We just need to document it properly, ideally with some hints how to configure Git to be compatible with every option.

**Comment by romani** [comment]
I am OK to make this Check , as I told mentioned before I will put approved label as we get out of moratorium period

**Comment by romani** [comment]
Enforcement of same line ends in whole repo might be  problematic if users keep scripts of different OS is same repo , example of problems when Linux line ends appear in windows scripts https://serverfault.com/a/429598

But Check is useful, and user need to configure it properly.

**Comment by jarek-przygodzki** [comment]
> This check is completely useless, as good VCS configuration will force local working copy to have Windows-like (CRLF) EOL-style on Windows clients and Linux EOL-style (LF) on Linux clients. In this way, the correctly configured clients should have multiple errors for checkstyle.

Concept of having your version control system be opinionated about line endings is rather new and controversial.

I certainly wouldn't say that every setup that does not do line ending normalization on checkout/checkin is  incorrect.

**Comment by romani** [comment]
@Brijeshthummar02 , can you help community to implement long awaiting Check ?

**Comment by Brijeshthummar02** [comment]
> [@Brijeshthummar02](https://github.com/Brijeshthummar02) , can you help community to implement long awaiting Check ?

Sure, can you just summarize it because there's multiple conversations and multiple PR referenced to this issue.

**Comment by romani** [comment]
@Brijeshthummar02 , 

please take changes from https://github.com/checkstyle/checkstyle/pull/3557 and send new PR, lets see how enforcing will work for our repsitory, as we do have windows files in our git.
If we can merge it, it might require some suppressions or usage of https://checkstyle.sourceforge.io/checks/regexp/regexpsingleline.html#fileExtensions , we can proceed with Check implementation.


example of Check that doing similar is https://checkstyle.sourceforge.io/checks/misc/newlineatendoffile.html#NewlineAtEndOfFile
attention to
https://checkstyle.sourceforge.io/property_types.html#LineSeparatorOption


Check can be named: LineEnding
```
<module name="Checker">
  <module name="LineEnding">
    <property name="lineSeparator" value="lf"/>
    <property name="fileExtensions" value="java, xml, cpp"/>
  </module>
</module>
```
and for windows:
```
<module name="Checker">
  <module name="LineEnding">
    <property name="lineSeparator" value="crlf"/>
    <property name="fileExtensions" value="bat, cmd"/>
  </module>
</module>
```

**Comment by Brijeshthummar02** [comment]
> [@Brijeshthummar02](https://github.com/Brijeshthummar02) ,
> 
> please take changes from [#3557](https://github.com/checkstyle/checkstyle/pull/3557) and send new PR, lets see how enforcing will work for our repsitory, as we do have windows files in our git. If we can merge it, it might require some suppressions or usage of https://checkstyle.sourceforge.io/checks/regexp/regexpsingleline.html#fileExtensions , we can proceed with Check implementation.
> 
> example of Check that doing similar is https://checkstyle.sourceforge.io/checks/misc/newlineatendoffile.html#NewlineAtEndOfFile attention to https://checkstyle.sourceforge.io/property_types.html#LineSeparatorOption
> 
> Check can be named: LineEnding
> 
> ```
> <module name="Checker">
>   <module name="LineEnding">
>     <property name="lineSeparator" value="lf"/>
>     <property name="fileExtensions" value="java, xml, cpp"/>
>   </module>
> </module>
> ```
> 
> and for windows:
> 
> ```
> <module name="Checker">
>   <module name="LineEnding">
>     <property name="lineSeparator" value="crlf"/>
>     <property name="fileExtensions" value="bat, cmd"/>
>   </module>
> </module>
> ```

let me see what can i do here.

**Comment by romani** [comment]
sample of how to do this by RegexpMultiline is merged to checkstyle-checks.xml

new module will be done in https://github.com/checkstyle/checkstyle/issues/18109
